File: internal/git/git.go
package git

import (
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/kawana77b/op/internal/util"
)

// 指定ディレクトリに.gitディレクトリがあるか見る。ない場合error
func DirExists(dir string) error {
	git := filepath.Join(dir, ".git")
	if !util.DirExists(git) {
		return errors.New("the .git directory does not exist")
	}

	return nil
}

// git config --get remote.origin.urlを実行し、その出力を返す
func GetOriginUrl(dir string) (string, error) {
	wd, _ := os.Getwd()
	defer func() {
		os.Chdir(wd)
	}()

	os.Chdir(dir)

	git, err := exec.LookPath("git")
	if err != nil {
		return "", errors.New("git is not installed")
	}

	args := []string{"config", "--get", "remote.origin.url"}
	buf, err := exec.Command(git, args...).Output()
	if err != nil {
		return "", errors.New("remote repository URL is not set")
	}

	url := string(buf)
	if len(url) == 0 {
		return "", errors.New("remote repository URL is not set")
	}

	return url, nil
}

// git@...のアドレスかどうか見る
func IsSSHAddr(addr string) bool {
	return regexp.MustCompile("^git@.+:.+.git$").MatchString(addr)
}

// .gitを取り除く
func TrimDotGit(url string) string {
	return regexp.MustCompile(".git$").ReplaceAllString(url, "")
}

// git sshアドレスからurl候補を生成
func SSHToUrls(ssh string) map[string]string {
	sshexp := regexp.MustCompile("^git@")

	domain := strings.Replace(sshexp.ReplaceAllString(ssh, ""), ":", "/", 1)
	domain = TrimDotGit(domain)

	return map[string]string{
		"https": "https://" + domain,
		"http":  "http://" + domain,
	}
}

File: src/open/git.rs
use std::{env, path::Path, vec};

use anyhow::{anyhow, Error, Result};
use regex::Regex;

use crate::shared::shutil;

fn check_if_git_directory(path: &str) -> Result<()> {
    let git_dir = Path::new(path).join(".git");
    match git_dir.is_dir() {
        true => Ok(()),
        false => Err(anyhow!("Not a git project directory")),
    }
}

fn check_git_availability() -> Result<()> {
    match shutil::look_path("git") {
        Some(_) => Ok(()),
        None => Err(anyhow!("git is not installed")),
    }
}

fn is_ssh(url: &str) -> bool {
    let re = Regex::new(r"^git@.+:.+.git$").unwrap();
    re.is_match(url)
}

fn get_cmd_output(cmd: &str, args: &[&str]) -> Result<String, Error> {
    let exec_result = shutil::run_subprocess_ex(cmd, args).and_then(|output| {
        String::from_utf8(output.stdout).or(Err(anyhow!("Failed to get command result")))
    });
    let raw = exec_result?;
    Ok(raw)
}

/// Format the url to like https://github.com. If it is ssh, convert it to https.
fn format_url(url: &str) -> String {
    let mut u = String::from(url.trim());
    if is_ssh(u.as_str()) {
        // git@github.com:xxxxxx/xxxxxxxxxx.git
        let sshexp = Regex::new(r"^git@").unwrap();
        u = sshexp.replace_all(&u, "").replacen(':', "/", 1);
        u = format!("https://{}", u);
    }
    // Remove the xxxxxx.git and trim
    u.trim_end_matches(".git").to_string()
}

/// Get the remote url of the git repository.
pub fn get_remote_url(dir: &str) -> Result<String> {
    if env::set_current_dir(dir).is_err() {
        return Err(anyhow!("Failed to change directory"));
    }
    check_if_git_directory(dir)?;
    check_git_availability()?;
    let output = get_cmd_output("git", &vec!["config", "--get", "remote.origin.url"])?;
    let trimmed = output.trim();
    if trimmed.len() == 0 {
        return Err(anyhow!("Failed to get remote url"));
    }
    Ok(format_url(trimmed))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_ssh() {
        assert_eq!(is_ssh("git@github.com:foo/bar.git"), true);
        assert_eq!(is_ssh("git@github.com:foo/bar"), false);
    }

    #[test]
    fn test_format_url() {
        let ssh_url = "git@github.com:foo/bar.git";
        let ssh_formatted = format_url(ssh_url);
        assert_eq!("https://github.com/foo/bar", ssh_formatted);

        let http_url = "https://github.com/foo/bar.git";
        let http_formatted = format_url(http_url);
        assert_eq!("https://github.com/foo/bar", http_formatted);
    }
}
